File: leetcode-cn/0001.0_Two_Sum.py
'''
apporach: hashmap
Time: O(N)
Space: O(N)

执行结果：通过
显示详情
执行用时：20 ms, 在所有 Python 提交中击败了76.20%的用户
内存消耗：
12.9 MB, 在所有 Python 提交中击败了90.98%的用户
'''


class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        need = {}
        size = len(nums)
        for i in range(size):
            if nums[i] in need:
                return [i, need[nums[i]]]
            need[target - nums[i]] = i

'''
hash table
T: O(N)
S: O(N)

执行用时：44 ms, 在所有 Python3 提交中击败了59.06% 的用户
内存消耗：16 MB, 在所有 Python3 提交中击败了27.18% 的用户
通过测试用例：57 / 57
'''
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num2idx = dict()
        for i, x in enumerate(nums):
            if target - x in num2idx:
                return [i, num2idx[target - x]]
            num2idx[x] = i 
        return []

File: cpp/0001.0_Two_Sum.cpp
/*
执行用时：8 ms, 在所有 C++ 提交中击败了92.05% 的用户
内存消耗：10.5 MB, 在所有 C++ 提交中击败了38.27% 的用户
通过测试用例：57 / 57
*/
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> um;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            if (um.count(target - nums[i]) > 0) {
                auto it = um.find(target - nums[i]);
                return {it->second, i};
            }
            um.insert({nums[i], i});
        }
        return {-1, -1};
    }
};


/*
执行用时：12 ms, 在所有 C++ 提交中击败了64.55% 的用户
内存消耗：10.4 MB, 在所有 C++ 提交中击败了52.98% 的用户
通过测试用例：57 / 57
*/
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::unordered_map<int, int> um;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            auto it = um.find(target - nums[i]);
            if (it != um.end()) {
                return {it->second, i};
            }
            um.insert(pair<int, int>(nums[i], i));
        }
        return {};
    }
};

